sdk/keyvault/azure_security_keyvault_keys/README.md

```diff
@@ -1,66 +1,370 @@
 
+### KeyBundle
+
+Azure Key Vault supports multiple key types and algorithms, and enables the use of hardware security modules (HSM) for high value keys.
+
 ### KeyClient
 
-The `KeyClient` struct provides methods to manage keys in the Azure Key Vault. You can create, retrieve, update, and delete keys, as well as perform cryptographic operations.
+The `KeyClient` provides asynchronous operations for working with Key Vault keys.
+
+### Thread safety
+
+We guarantee that all client instance methods are thread-safe and independent of each other. This ensures that the recommendation of reusing client instances is always safe, even across threads.
 
 ## Examples
 
-### Creating a key
+The following section provides several code snippets using the `KeyClient`, covering some of the most common Azure Key Vault keys service related tasks:
+
+* [Create a key](#create-a-key)
+* [Retrieve a key](#retrieve-a-key)
+* [Update an existing key](#update-an-existing-key)
+* [Delete a key](#delete-a-key)
+* [List keys](#list-keys)
+
+### Create a key
+
+`create_key` creates a Key Vault key to be stored in the Azure Key Vault. If a key with the same name already exists, then a new version of the key is created.
 
 ```rust no_run
 use azure_identity::DefaultAzureCredential;
-use azure_security_keyvault_keys::{models::KeyCreateParameters, KeyClient};
+use azure_security_keyvault_keys::{
+    models::{KeyBundle, KeyCreateParameters, JsonWebKeyCurveName, JsonWebKeyType},
+    ResourceExt, KeyClient,
+};
 
 #[tokio::main]
 async fn main() -> Result<(), Box<dyn std::error::Error>> {
     let credential = DefaultAzureCredential::new()?;
+    let client = KeyClient::new(
+        "https://your-key-vault-name.vault.azure.net/",
+        credential.clone(),
+        None,
+    )?;
+
+    // Create an EC key.
+    let body = KeyCreateParameters {
+        kty: Some(JsonWebKeyType::EC),
+        curve: Some(JsonWebKeyCurveName::P256),
+        ..Default::default()
+    };
+
+    let key = client
+        .create_key("key-name", body.try_into()?, None)
+        .await?
+        .into_body()
+        .await?;
 
+    println!(
+        "Key Name: {:?}, Type: {:?}, Version: {:?}",
+        key.resource_id()?.name,
+        key.key.as_ref().map(|k| k.kty.as_ref()),
+        key.resource_id()?.version,
+    );
+
+    Ok(())
+}
+```
+
+### Retrieve a key
+
+`get_key` retrieves a public key (or only metadata for symmetric keys) previously stored in the Azure Key Vault. Setting the `key-version` to an empty string will return the latest version.
+
+```rust no_run
+use azure_identity::DefaultAzureCredential;
+use azure_security_keyvault_keys::{models::KeyBundle, KeyClient};
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let credential = DefaultAzureCredential::new()?;
     let client = KeyClient::new(
-        "https://<your-key-vault-name>.vault.azure.net",
+        "https://your-key-vault-name.vault.azure.net/",
         credential.clone(),
```

Comment:
QQ: You've mentioned that best practices is to require the caller to `clone()` the credentials.

Does that imply that the service client should *not* clone the credentials since we're expecting the caller to `clone()` before passing the credentials into the service client? Otherwise, how do we avoid double cloning the credentials?

Or is this to safeguard against the case where the caller is going to be passing the same credentials client into multiple service clients? In which case, wouldn't the borrow checker guide them to cloning?

Is it better to say that double cloning on an arc is a best practice, or to let the borrow checker fail when someone accidentally consumes the arc?

Reply:
I'd rather the customer do it so we're not needless. Let the borrow checking complain if necessary.

Reply:
To confirm: we should NOT clone the credentials when saving them in the service client object and instead we should consume the credential object? Not a huge deal, but I would like to be consistent across the SDKs.